File: C++/BFS/ZigZagTraversal.cpp
/*
*   @author
*   Aakash Verma
*
*   Output:
*   1
*   3 2
*   4 5 6 7
*/

#include <bits/stdc++.h> 
using namespace std;

struct Node {
    int data;
    struct Node *left, *right;
    Node(int data) { 
        this->data = data; 
        left = right = NULL; 
    }
};

struct Node *root = NULL;

vector<vector<int>> zigZagTraversal(struct Node *root) {
    vector<vector<int>> bfs;
    bool leftToRight = true;
    if(root == NULL)
        return bfs;
    queue<Node *> queue;
    queue.push(root);
    while(!queue.empty()) {
        int levelSize = queue.size();
        vector<int> currentLevel(levelSize);
        for(int i = 0; i < levelSize; i++) {
            Node *current = queue.front();
            queue.pop();
            
            if(leftToRight)
                currentLevel[i] = current->data;
            else
                currentLevel[levelSize - i - 1] = current->data;

            if(current->left != NULL)
                queue.push(current->left);
            if(current->right != NULL)
                queue.push(current->right);
        }
        bfs.push_back(currentLevel);
        leftToRight = !leftToRight;
    }
    return bfs;
}
    
int main(int argc, char const *argv[]) {
    root = new Node(1);
    root->left = new Node(2);
    root->right = new Node(3);
    root->left->left = new Node(4);
    root->left->right = new Node(5);
    root->right->left = new Node(6);
    root->right->right = new Node(7);
    vector<vector<int>> ans = zigZagTraversal(root);
    for(auto level : ans) {
        for(auto num : level) {
            cout << num << " ";
        }
        cout << endl;
    }
}

File: Python/BFS/ZigZagTraversal.py
# @author
# Aakash Verma

# Output:
# [1]
# [3, 2]
# [4, 5, 6, 7] 

from collections import *

class Node:
    def __init__(self, data):
        self.data = data
        self.right = self.left = None

class ZigZagTraversal:
    
    def __init__(self): 
        self.root = None

    def traverse(self, root):
        bfs = []
        left_to_right = True
        if root is None:
            return bfs
        queue = deque([])
        queue.append(self.root)
        while len(queue) != 0:
            level_size = len(queue)
            current_level = deque()
            for i in range(level_size):
                current = queue.popleft()
                
                if left_to_right:
                    current_level.append(current.data)
                else:
                    current_level.appendleft(current.data)

                if current.left is not None:
                    queue.append(current.left)
                if current.right is not None:
                    queue.append(current.right)
            bfs.append(current_level)
            left_to_right = not left_to_right
        return bfs
    
if __name__ == '__main__':
    tree = ZigZagTraversal()
    tree.root = Node(1)
    tree.root.left = Node(2)
    tree.root.right = Node(3)
    tree.root.left.left = Node(4)
    tree.root.left.right = Node(5)
    tree.root.right.left = Node(6)
    tree.root.right.right = Node(7)
    ans = tree.traverse(tree.root)
    for level in ans:
        print(list(level))